students/nam_vo/lesson04/mailroom_part2.py

```diff
@@ -0,0 +1,151 @@
+}
+
+def send_single():
+    """Compose an email thanking the donor for their generous donation."""
+    # Ask for donor's full name
+    name = input("Enter donor's full name (or 'list' to list all donor' names): ")
+    # Continue asking for donor's full name
+    while name == "list":
+        # Display list of donor names
+        print_donor(donor_dict)
+        # Ask for donor's full name
+        name = input("Enter donor's full name (or 'list' to list all donor' names): ")
+    # Add donor to dictionary if it's new
+    donor_dict.setdefault(name, [])
+    # Prompt for a donation amount
+    response = input("Enter donation amount: $")
+    # Convert user's prompt in string to float
+    amount = float(response)
+    # Add new amount to the donation history of the selected donor
+    donor_dict[name].append(amount)
+    # Compose thank you email to the donor
+    prompt = "\nDear {:s},\n\n        Thank you for your very kind donation of ${:.2f}.\n\n        It will be put to very good use.\n\n                       Sincerely,\n                          -The Team\n".format(name, amount)
+    print(prompt)
+    
+def create_report():
+    """
+    Print a list of donors, sorted by total historical donation amount.
+
+    Sample:
+    Donor Name                | Total Given | Num Gifts | Average Gift
+    ------------------------------------------------------------------
+    William Gates, III         $  653784.49           2  $   326892.24
+    Mark Zuckerberg            $   16396.10           3  $     5465.37
+    Jeff Bezos                 $     877.33           1  $      877.33
+    Paul Allen                 $     708.42           3  $      236.14
+
+    """
+    # Sort the donor's dictionary by total historical donation amount
+    sorted_donor_list = sorted(donor_dict.items(), key=sort_key, reverse=True)
+    # Print header
+    fheader = "{:<26}|{:^13}|{:^11}|{:>13}"
+    print(fheader.format("Donor Name", "Total Given", "Num Gifts", "Average Gift"))
+    print('-' * 66)
+    # Get body format
+    fbody = "{:<27}${:>11.2f}{:>12}  ${:>12.2f}"
+    # Loop thru each donor in the dictionary
+    for donor in sorted_donor_list:
+        # Get total donated amount, number of gifts, and average gift amount
+        total_given = sum(donor[1])
+        num_gifts = len(donor[1])
+        ave_gift = total_given / num_gifts
+        # Print report
+        print(fbody.format(donor[0], total_given, num_gifts, ave_gift))
+
+def send_all():
+    """Generate a thank you letter for each donor in the list and write each letter to current working directory as text file"""
+    # Loop thru each donor's name in the dictionary
+    for (name, donations) in donor_dict.items():
+        # Get the latest amount of donation
+        amount = donations[len(donations) - 1]
+        # Assign file name
+        fileName = name.replace(' ', '_').replace(',', '') + ".txt"
+        # Open file for writing
+        with open(fileName, 'w') as outFile:
```

Comment:
pep8 violation: make sure to use snake_case style for var names -> `outFile` -> `out_file`